kore/src/Kore/Unification/UnifierImpl.hs

```diff
@@ -71,123 +71,160 @@ simplifyAnds
         )
     => NonEmpty (TermLike variable)
     -> unifier (Pattern variable)
-simplifyAnds patterns = do
-    let
-        simplifyAnds'
-            :: Pattern variable
-            -> TermLike variable
-            -> unifier (Pattern variable)
-        simplifyAnds' intermediate pat =
-            case Cofree.tailF (Recursive.project pat) of
-                AndF And { andFirst = lhs, andSecond = rhs } ->
-                    foldM simplifyAnds' intermediate [lhs, rhs]
-                _ -> do
-                    result <- termUnification (Pattern.term intermediate) pat
-                    predSubst <-
-                        mergePredicatesAndSubstitutionsExcept
-                            [ Pattern.predicate result
-                            , Pattern.predicate intermediate
-                            ]
-                            [ Pattern.substitution result
-                            , Pattern.substitution intermediate
-                            ]
-                    return Pattern.Conditional
-                        { term = Pattern.term result
-                        , predicate = Conditional.predicate predSubst
-                        , substitution = Conditional.substitution predSubst
-                        }
+simplifyAnds (NonEmpty.sort -> patterns) = do
     result <- foldM simplifyAnds' Pattern.top patterns
-    if Predicate.isFalse . Pattern.predicate $ result
-        then return Pattern.bottom
-        else return result
-
-groupSubstitutionByVariable
-    :: (Ord variable, SortedVariable variable)
-    => [(UnifiedVariable variable, TermLike variable)]
-    -> [[(UnifiedVariable variable, TermLike variable)]]
-groupSubstitutionByVariable =
-    groupBy ((==) `on` fst) . sortBy (compare `on` fst) . map sortRenaming
+    TopBottom.guardAgainstBottom result
+    return result
   where
-    sortRenaming (var, Var_ var') | var' < var = (var', mkVar var)
-    sortRenaming eq = eq
+    simplifyAnds'
+        :: Pattern variable
+        -> TermLike variable
+        -> unifier (Pattern variable)
+    simplifyAnds' intermediate pat =
+        case Cofree.tailF (Recursive.project pat) of
+            AndF And { andFirst = lhs, andSecond = rhs } ->
+                foldM simplifyAnds' intermediate [lhs, rhs]
+            _ -> do
+                result <- termUnification (Pattern.term intermediate) pat
+                normalizeExcept (result <* intermediate)
```

Comment:
splitting intermediate into its term and predicate (`Pattern.splitTerm`) and using `andCondition` (or `<>` if it works here) instead of `<*` would make it more clear that we're not dropping the `intermediate`'s term by mistake here.